# Викторина
в которой предполагается единственный верный ответ на каждый вопрос

In [169]:
import pandas as pd

In [170]:
data = pd.read_csv('movie_bd_v5.csv')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1889 entries, 0 to 1888
Data columns (total 14 columns):
imdb_id                 1889 non-null object
budget                  1889 non-null int64
revenue                 1889 non-null int64
original_title          1889 non-null object
cast                    1889 non-null object
director                1889 non-null object
tagline                 1889 non-null object
overview                1889 non-null object
runtime                 1889 non-null int64
genres                  1889 non-null object
production_companies    1889 non-null object
release_date            1889 non-null object
vote_average            1889 non-null float64
release_year            1889 non-null int64
dtypes: float64(1), int64(4), object(9)
memory usage: 206.7+ KB


In [171]:
len(data)

1889

In [172]:
data.sample(5)

imdb_id     budget    revenue                   original_title  \
885   tt0283111    5000000   38275483  National Lampoon’s Van Wilder   
177   tt2582846   12000000  305151265           The Fault in Our Stars   
1423  tt0410097    8000000   22201636                    Hustle & Flow   
1575  tt0298148  150000000  919838758                          Shrek 2   
724   tt1318514   93000000  482860185   Rise of the Planet of the Apes   

                                                   cast  \
885   Ryan Reynolds|Tara Reid|Tim Matheson|Kal Penn|...   
177   Shailene Woodley|Ansel Elgort|Nat Wolff|Laura ...   
1423  Terrence Howard|Anthony Anderson|Taryn Manning...   
1575  Mike Myers|Eddie Murphy|Cameron Diaz|Julie And...   
724   Andy Serkis|Karin Konoval|Freida Pinto|James F...   

                                       director  \
885                                 Walt Becker   
177                                  Josh Boone   
1423                               Craig Brewer   
1575  Andrew Adamson|Kelly Asbury|Conrad Vernon   
724                                Rupert Wyatt   

                                                tagline  \
885                          Don't Graduate. Celebrate.   
177                                 One Sick Love Story   
1423  The music will inspire them. The dream will un...   
1575                          Once upon another time...   
724                       Evolution Becomes Revolution.   

                                               overview  runtime  \
885   Van Wilder is an outgoing, and extremely popul...       92   
177   Despite the tumor-shrinking medical miracle th...      125   
1423  With help from his friends, a Memphis pimp in ...      116   
1575  Shrek, Fiona and Donkey set off to Far, Far Aw...       93   
724   Scientist Will Rodman is determined to find a ...      105   

                                         genres  \
885                              Comedy|Romance   
177                               Romance|Drama   
1423                                      Drama   
1575  Adventure|Animation|Comedy|Family|Fantasy   
724       Thriller|Action|Drama|Science Fiction   

                                   production_companies release_date  \
885                Tapestry Films|Artisan Entertainment    3/29/2002   
177   Fox 2000 Pictures|Temple Hill Entertainment|TS...    5/16/2014   
1423                                          MTV Films    7/22/2005   
1575  DreamWorks SKG|Pacific Data Images (PDI)|Dream...    5/19/2004   
724                                    20th Century Fox     8/3/2011   

      vote_average  release_year  
885            5.9          2002  
177            7.8          2014  
1423           6.2          2005  
1575           6.5          2004  
724            6.9          2011

# Предобработка датасета

In [173]:
answers = {}  # ответы на вопросы теста

data = data.assign(
    release_date=lambda row: pd.to_datetime(row.release_date),
    profit=lambda row: row.revenue - row.budget)


In [174]:
display(data)

imdb_id     budget     revenue                        original_title  \
0     tt0369610  150000000  1513528810                        Jurassic World   
1     tt1392190  150000000   378436354                    Mad Max: Fury Road   
2     tt2908446  110000000   295238201                             Insurgent   
3     tt2488496  200000000  2068178225          Star Wars: The Force Awakens   
4     tt2820852  190000000  1506249360                             Furious 7   
...         ...        ...         ...                                   ...   
1884  tt0120903   75000000   157299717                                 X-Men   
1885  tt0192255   22000000    13555988                    The Little Vampire   
1886  tt0131704   76000000    35134820  The Adventures of Rocky & Bullwinkle   
1887  tt0162983   40000000    36037909                            Hanging Up   
1888  tt0163676   15000000     5217498                          The In Crowd   

                                                   cast          director  \
0     Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1     Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2     Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   
3     Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...       J.J. Abrams   
4     Vin Diesel|Paul Walker|Jason Statham|Michelle ...         James Wan   
...                                                 ...               ...   
1884  Patrick Stewart|Hugh Jackman|Ian McKellen|Hall...      Bryan Singer   
1885  Richard E. Grant|Jonathan Lipnicki|Jim Carter|...          Uli Edel   
1886  Rene Russo|Jason Alexander|Piper Perabo|Randy ...       Des McAnuff   
1887  Meg Ryan|Diane Keaton|Lisa Kudrow|Walter Matth...      Diane Keaton   
1888  Susan Ward|Lori Heuring|Matthew Settle|Nathan ...      Mary Lambert   

                                                tagline  \
0                                     The park is open.   
1                                    What a Lovely Day.   
2                            One Choice Can Destroy You   
3                         Every generation has a story.   
4                                   Vengeance Hits Home   
...                                                 ...   
1884                                   Evolution Begins   
1885  They're not just best friends, they're blood b...   
1886            This summer it's not the same old bull.   
1887                   Every family has a few hang-ups.   
1888                       What would you do to get in?   

                                               overview  runtime  \
0     Twenty-two years after the events of Jurassic ...      124   
1     An apocalyptic story set in the furthest reach...      120   
2     Beatrice Prior must confront her inner demons ...      119   
3     Thirty years after defeating the Galactic Empi...      136   
4     Deckard Shaw seeks revenge against Dominic Tor...      137   
...                                                 ...      ...   
1884  Two mutants, Rogue and Wolverine, come to a pr...      104   
1885  Based on the popular books, the story tells of...       95   
1886  Rocky and Bullwinkle have been living off the ...       88   
1887  A trio of sisters bond over their ambivalence ...       94   
1888  A mentally disturbed young woman takes a job a...      105   

                                         genres  \
0     Action|Adventure|Science Fiction|Thriller   
1     Action|Adventure|Science Fiction|Thriller   
2            Adventure|Science Fiction|Thriller   
3      Action|Adventure|Science Fiction|Fantasy   
4                         Action|Crime|Thriller   
...                                         ...   
1884           Adventure|Action|Science Fiction   
1885                      Horror|Family|Foreign   
1886   Adventure|Animation|Action|Comedy|Family   
1887                               Comedy|Drama   
1888                               

# 1. У какого фильма из списка самый большой бюджет?
Варианты ответов:
1. The Dark Knight Rises (tt1345836)
2. Spider-Man 3 (tt0413300)
3. Avengers: Age of Ultron (tt2395427)
4. The Warrior's Way	(tt1032751)
5. Pirates of the Caribbean: On Stranger Tides (tt1298650)

In [175]:
data[data['budget'] == data['budget'].max()][['imdb_id', 'original_title']]

imdb_id                               original_title
723  tt1298650  Pirates of the Caribbean: On Stranger Tides

In [176]:
answers[1] = "Pirates of the Caribbean: On Stranger Tides"

# 2. Какой из фильмов самый длительный (в минутах)
1. The Lord of the Rings: The Return of the King	(tt0167260)
2. Gods and Generals	(tt0279111)
3. King Kong	(tt0360717)
4. Pearl Harbor	(tt0213149)
5. Alexander	(tt0346491)

In [177]:
data[data['runtime'] == data['runtime'].max()][['imdb_id', 'original_title']]

imdb_id     original_title
1157  tt0279111  Gods and Generals

In [178]:
answers[2] = "Gods and Generals"

# 3. Какой из фильмов самый короткий (в минутах)
Варианты ответов:

1. Home on the Range	tt0299172
2. The Jungle Book 2	tt0283426
3. Winnie the Pooh	tt1449283
4. Corpse Bride	tt0121164
5. Hoodwinked!	tt0443536

In [179]:
data[data['runtime'] == data['runtime'].min()][['imdb_id', 'original_title']]

imdb_id   original_title
768  tt1449283  Winnie the Pooh

In [180]:
answers[3] = "Winnie the Pooh"

# 4. Средняя длительность фильма?

Варианты ответов:
1. 115
2. 110
3. 105
4. 120
5. 100


In [181]:
time = data.runtime.mean()
print(round(time))

110


In [182]:
answers[4] = 110

# 5. Средняя длительность фильма по медиане?
Варианты ответов:
1. 107
2. 112
3. 101
4. 120
5. 115




In [183]:
time = data.runtime.median()
print(time)

107.0


In [184]:
answers[5] = 107

# 6. Какой самый прибыльный фильм?
Варианты ответов:
1. The Avengers	tt0848228
2. Minions	tt2293640
3. Star Wars: The Force Awakens	tt2488496
4. Furious 7	tt2820852
5. Avatar	tt0499549

In [185]:
data[data['profit'] == data['profit'].max()][['imdb_id', 'original_title']]

imdb_id original_title
239  tt0499549         Avatar

In [186]:
answers[6] = "Avatar"

# 7. Какой фильм самый убыточный?
Варианты ответов:
1. Supernova tt0134983
2. Frozen tt2294629
3. Flushed Away	tt0424095
4. The Adventures of Pluto Nash	tt0180052
5. The Lone Ranger	tt1210819

In [187]:
data[data['profit'] == data['profit'].min()][['imdb_id', 'original_title']]

imdb_id   original_title
1245  tt1210819  The Lone Ranger

In [188]:
answers[7] = "The Lone Ranger"

# 8.У скольких фильмов из датасета объем сборов оказался выше бюджета?
Варианты ответов:
1. 1478
2. 1520
3. 1241
4. 1135
5. 1398


In [189]:
len(data[data['profit'] > 0])

1478

In [190]:
answers[8] = 1478

# 9. Самый прибыльный фильм в 2008 году?
Варианты ответов:
1. Madagascar: Escape 2 Africa	tt0479952
2. Iron Man	tt0371746
3. Kung Fu Panda	tt0441773
4. The Dark Knight	tt0468569
5. Mamma Mia!	tt0795421

In [191]:
most_profit_2008 = data.query('release_year == 2008').query('profit == profit.max()')
most_profit_2008[['imdb_id', 'original_title']]

imdb_id   original_title
599  tt0468569  The Dark Knight

In [192]:
answers[9] = "The Dark Knight"

# 10. Самый убыточный фильм за период с 2012 по 2014 (включительно)?
Варианты ответов:
1. Winter's Tale	tt1837709
2. Stolen	tt1656186
3. Broken City	tt1235522
4. Upside Down	tt1374992
5. The Lone Ranger	tt1210819


In [193]:
least_profit_2012_2014 = data.query('release_year in [2012,2013,2014]').query('profit == profit.min()')
least_profit_2012_2014[['imdb_id', 'original_title']]

imdb_id   original_title
1245  tt1210819  The Lone Ranger

In [194]:
answers[10] = "The Lone Ranger"

# 11. Какого жанра фильмов больше всего?
Варианты ответов:
1. Action
2. Adventure
3. Drama
4. Comedy
5. Thriller

In [195]:
from collections import Counter

In [196]:
genres = Counter()
for row in data.itertuples():
    genres.update(row.genres.split('|'))
genres.most_common(1)

[('Drama', 782)]

In [197]:
answers[11] = "Drama"

# 12. Какого жанра среди прибыльных фильмов больше всего?
Варианты ответов:
1. Drama
2. Comedy
3. Action
4. Thriller
5. Adventure

In [198]:
genres = Counter()
for row in data[data.profit > 0].itertuples():
    genres.update(row.genres.split('|'))
genres.most_common(1)

[('Drama', 560)]

In [199]:
answers[12] = "Drama"

# 13.  У какого режиссёра самые большие суммарные кассовые сборы?
Варианты ответов:
1. Steven Spielberg
2. Christopher Nolan
3. David Yates
4. James Cameron
5. Peter Jackson

In [200]:
directors = data.assign(director=data.director.str.split('|')).explode('director')
grouped_directors = directors.groupby('director')

def sum_revenue(series):
    return series.sum()

grouped_directors.revenue.agg(sum_revenue).idxmax()

'Peter Jackson'

In [201]:
answers[13] = "Peter Jackson"

# 14. Какой режиссер снял больше всего фильмов в стиле Action?
Варианты ответов:
1. Ridley Scott
2. Guy Ritchie
3. Robert Rodriguez
4. Quentin Tarantino
5. Tony Scott

In [202]:
actions = data[data['genres'].str.contains('Action')]
display(actions['director'].value_counts())

Paul W.S. Anderson    7
Michael Bay           7
Robert Rodriguez      6
Ridley Scott          6
Antoine Fuqua         6
                     ..
Steven Brill          1
Alexander Witt        1
Peter Segal           1
Jay Roach             1
Roger Michell         1
Name: director, Length: 345, dtype: int64

In [203]:
answers[14] = "Robert Rodriguez"

# 15. Фильмы с каким актером принесли самые высокие кассовые сборы в 2012 году?
Варианты ответов:
1. Nicolas Cage
2. Tom Hardy
3. Chris Hemsworth
4. Jim Sturgess
5. Emma Stone


In [204]:
films_2012 = data[data['release_year'] == 2012]
films_2012_max_revenue = films_2012[films_2012['revenue'] == films_2012['revenue'].max()]
display(films_2012_max_revenue['cast'].value_counts())

Robert Downey Jr.|Chris Evans|Mark Ruffalo|Chris Hemsworth|Scarlett Johansson    1
Name: cast, dtype: int64

In [205]:
answers[15] = "Chris Hemsworth"

# 16. Какой актер снялся в большем количестве высокобюджетных фильмов? Примечание: в фильмах, где бюджет выше среднего по данной выборке
Варианты ответов:
1. Tom Cruise
2. Mark Wahlberg 
3. Matt Damon
4. Angelina Jolie
5. Adam Sandler

In [206]:
actors = data.assign(cast=data.cast.str.split('|')).explode('cast')
actors[actors.budget >= actors.budget.mean()].groupby('cast').size().idxmax()

'Matt Damon'

In [207]:
answers[16] = "Matt Damon"

# 17. В фильмах какого жанра больше всего снимался Nicolas Cage?  
Варианты ответа:
1. Drama
2. Action
3. Thriller
4. Adventure
5. Crime

In [208]:
genres = Counter()
for row in actors[actors.cast == 'Nicolas Cage'].itertuples():
    genres.update(row.genres.split('|'))
genres.most_common(1)

[('Action', 17)]

In [209]:
answers[19] = "Action"

# 18.  Самый убыточный фильм от Paramount Pictures?
Варианты ответа:
1. K-19: The Widowmaker tt0267626
2. Next tt0435705
3. Twisted tt0315297
4. The Love Guru tt0811138
5. The Fighter tt0964517

In [210]:
worst_paramount = data[data['production_companies'].str.contains("Paramount Pictures")]
worst_paramount = worst_paramount[worst_paramount['profit'] == worst_paramount['profit'].min()]
worst_paramount[['imdb_id', 'original_title']]

imdb_id        original_title
925  tt0267626  K-19: The Widowmaker

In [211]:
answers[18] = "K-19: The Widowmaker"

# 19. Какой год стал самым успешным по суммарным кассовым сборам?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2002
5. 2015

In [212]:
year_groups = data.groupby('release_year')[['revenue']].sum()
year_groups[year_groups.revenue == year_groups.revenue.max()]

revenue
release_year             
2015          25449202382

In [213]:
answers[19] = 2015

# 20. Какой Самый прибыльный год для студии Warner Bros?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2010
5. 2015

In [214]:
warner_films = data[data.production_companies.str.contains('Warner Bros')]

year_groups = warner_films.groupby('release_year')[['profit']].sum()
year_groups[year_groups.profit == year_groups.profit.max()]

profit
release_year            
2014          2295464519

In [215]:
answers[20] = 2014

# 21. В каком месяце за все годы суммарно вышло больше всего фильмов?
Варианты ответа:
1. Январь 
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [216]:
from calendar import month_name

In [217]:
month = data.groupby(data.release_date.dt.month).size().idxmax()
month_name[month]

'September'

In [218]:
answers[21] = "September"

# 22. Сколько суммарно вышло фильмов летом (за июнь, июль, август)?
Варианты ответа:

1. 345
2. 450
3. 478
4. 523
5. 381

In [219]:
len(data[data.release_date.dt.month.between(6, 8)])

450

In [220]:
answers[22] = 450

# 23.  Для какого режиссера зима — самое продуктивное время года?
Варианты ответа:
1. Steven Soderbergh
2. Christopher Nolan
3. Clint Eastwood
4. Ridley Scott
5. Peter Jackson

In [221]:
# для упрощения нумерации используем деление по модулю 12
sample = data[data.release_date.dt.month.mod(12) <= 2]
sample.groupby('director').size().idxmax()

'Peter Jackson'

In [222]:
answers[23] = "Peter Jackson"

# 24. Названия фильмов какой студии в среднем самые длинные по количеству символов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [223]:
sample = data.assign(title_length=data.original_title.str.len())
sample.groupby('production_companies').title_length.mean().idxmax()

'Twentieth Century Fox Film Corporation|Four By Two Productions'

In [224]:
answers[24] = "Four By Two Productions"

# 25.Описания фильмов какой студии в среднем самые длинные по количеству слов?
Варианты ответа:
1. Universal Pictures 
2. Warner Bros
3. Midnight Picture Show
4. Paramount Pictures
5. Total Entertainment
6. Four By Two Productions (?)

In [225]:
sample = data.assign(title_length=data.original_title.str.split().str.len())
display(sample.groupby('production_companies').title_length.mean().idxmax())

'Twentieth Century Fox Film Corporation|Four By Two Productions'

In [226]:
answers[25] = "Four By Two Productions"

# 26. Какие фильмы входят в 1 процент лучших по рейтингу?
Варианты ответа:
1. Inside Out, The Dark Knight, 12 Years a Slave
2. BloodRayne, The Adventures of Rocky & Bullwinkle
3. The Lord of the Rings: The Return of the King, Upside Down
4. 300, Lucky Number Slevin, Kill Bill: Vol. 1
5. Upside Down, Inside Out, Iron Man

In [227]:
from functools import partial

In [228]:
one_percent = partial(pd.Series.quantile, q=0.99)
one_percent.__name__ = '1%'

In [229]:
best_titles = data[data.vote_average > data.vote_average.agg(one_percent)].original_title

In [230]:
candidates = [
    ['Inside Out','The Dark Knight', '12 Years a Slave'],
    ['BloodRayne', 'The Adventures of Rocky & Bullwinkle'],
    ['The Lord of the Rings: The Return of the King', 'Upside Down'],
    ['300', 'Lucky Number Slevin', 'Kill Bill: Vol. 1'],
    ['Upside Down', 'Inside Out', 'Iron Man']
]

[titles for titles in candidates if set(titles).issubset(best_titles)]

[['Inside Out', 'The Dark Knight', '12 Years a Slave']]

In [231]:
answers[26] = ["Inside Out", "The Dark Knight", "12 Years a Slave"]

# 27. Какие актеры чаще всего снимаются в одном фильме вместе
Варианты ответа:
1. Johnny Depp & Helena Bonham Carter
2. Ben Stiller & Owen Wilson
3. Vin Diesel & Paul Walker
4. Adam Sandler & Kevin James
5. Daniel Radcliffe & Rupert Grint

In [232]:
from itertools import combinations
from collections import Counter

In [233]:
actors_pairs = Counter()
for row in data.itertuples():
    actors_pairs.update(combinations(row.cast.split('|'), 2))
actors_pairs.most_common(1)

[(('Daniel Radcliffe', 'Rupert Grint'), 8)]

In [234]:
answers[27] = ["Daniel Radcliffe", "Rupert Grint"]

# Submission

In [235]:
len(answers)

26

In [236]:
pd.DataFrame.from_dict(answers, orient='index', columns=['Answer'])

Answer
1       Pirates of the Caribbean: On Stranger Tides
2                                 Gods and Generals
3                                   Winnie the Pooh
4                                               110
5                                               107
6                                            Avatar
7                                   The Lone Ranger
8                                              1478
9                                   The Dark Knight
10                                  The Lone Ranger
11                                            Drama
12                                            Drama
13                                    Peter Jackson
14                                 Robert Rodriguez
15                                  Chris Hemsworth
16                                       Matt Damon
19                                             2015
18                             K-19: The Widowmaker
20                                             2014
21                                        September
22                                              450
23                                    Peter Jackson
24                          Four By Two Productions
25                          Four By Two Productions
26  [Inside Out, The Dark Knight, 12 Years a Slave]
27                 [Daniel Radcliffe, Rupert Grint]